ldms/src/sampler/ibmad_sampler/ibmad_sampler.c

```diff
@@ -627,74 +620,66 @@ static void dump_port_filters()
 	}
 }
 
-static int parse_port_filters(const char *include, const char *exclude)
+static int parse_port_filters(const char *val)
 {
-	if (include && exclude) {
-                log_fn(LDMSD_LERROR, SAMP": config: specify either include or exclude option but not both.\n");
-		return 1;
-	}
-	const char *val = NULL;
-	if (include) {
-		val = include;
-		conf.port_filter = PORT_FILTER_INCLUDE;
-	}
-	if (exclude) {
-		val = exclude;
-		conf.port_filter = PORT_FILTER_EXCLUDE;
-	}
 	int k = 0;
 	int num_ca = 0;
-	if (val) {
-		uint64_t num;
-		size_t len = strlen(val);
-		char s[len+1];
-		strcpy(s, val);
-		char *pch, *saveptr;
-		pch = strtok_r(s, ",", &saveptr);
-		while (pch != NULL){
-			char *dot = strchr(pch, '.');
-			num = 0;
-			if (dot) {
-				char *end = NULL;
-				dot[0] = '\0';
-				dot++;
-				errno = 0;
-				num = strtoull(dot, &end, 10);
-				if (*end != '\0' || errno == ERANGE) {
-					log_fn(LDMSD_LERROR, SAMP": config: "
-						"%s port invalid: %s.\n",
-						val, dot);
-				}
-				if (num > 63) {
-					log_fn(LDMSD_LERROR, SAMP": config: "
-						"%s port > 63: %" PRIu64 ".\n",
-						val, num);
-					return 1;
-				}
+	if (!val) {
+		return 0;
+	}
+	unsigned long num;
+	size_t len = strlen(val);
+	char s[len+1];
+	strcpy(s, val);
+	char *pch, *saveptr;
+	pch = strtok_r(s, ",", &saveptr);
+	while (pch != NULL){
+		char *dot = strchr(pch, '.');
+		num = 0;
+		if (dot) {
+			char *end = NULL;
+			dot[0] = '\0';
+			dot++;
+			errno = 0;
+			num = strtoul(dot, &end, 10);
+			if (*end != '\0' || errno == ERANGE) {
+				log_fn(LDMSD_LERROR, SAMP": config: "
+					"%s port invalid: %s.\n",
+					val, dot);
+				return 1;
 			}
-			for (k = 0; k < num_ca; k++) {
-				if (strcmp(conf.ports[k].ca_name, pch) == 0) {
-					conf.ports[k].port_bits |= (1 << num);
-					break;
-				}
+			if (num > 63) {
+				log_fn(LDMSD_LERROR, SAMP": config: "
+					"%s port > 63: %lu.\n",
+					val, num);
+				return 1;
 			}
-			if (k == num_ca) {
-				if (k > MAX_CA_NAMES) {
-					log_fn(LDMSD_LERROR, SAMP": config: "
-						"too many CA in %s\n", val);
-					return 1;
-				}
-				strcpy(conf.ports[k].ca_name, pch);
+		}
+		for (k = 0; k < num_ca; k++) {
+			if (strcmp(conf.ports[k].ca_name, pch) == 0) {
 				conf.ports[k].port_bits |= (1 << num);
```

Comment:
Setting the wrong value here when a port isn't set and then fixing it later doesn't make the code as readable as it could be. This is where you should set ports_bits = UINT64_MAX.

Reply:
Rather than being a 'bug', it is just that lines 660 and 671 could both be moved to (insert@) line 675. This would leave us with port_bits being updated once for each port parsed. The setting of bit 0 (the wildcard bit, since there is never an IB port 0) and then resetting the entire port_bits to uint64_max leads to an extra assignment in the wildcard case (but not a bug).